### PR TITLE
Add controlled gate support to `lightning.tensor`

### PR DESCRIPTION
**Pull request by multiphaseCFD**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

Add controlled gate support to `lightning.tensor`. This could improve the performance of `lightning.tensor` by avoiding  full gate matrix build for controlled gates. However, as of v24.08, cutensornet only supports controlled gates with only 1 wire target. [SC-72522]

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/TNCudaBase.hpp [hunk 1]
```diff
@@ -244,6 +246,87 @@ class TNCudaBase : public TensornetBase<PrecisionT, Derived> {
...
+                             bool adjoint = false,
+                             const std::vector<PrecisionT> &params = {0.0},
+                             const std::vector<ComplexT> &gate_matrix = {}) {
+        // TODO: Need to revisit this line of code once `cutensornet` supports
+        // multi-target wire controlled gates
+        PL_ABORT_IF_NOT(targetWires.size() == 1,
+                        "Unsupported controlled gate: cutensornet only "
+                        "supports 1-wire target controlled gates");
+
+        auto &&par = (params.empty()) ? std::vector<PrecisionT>{0.0} : params;
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/TNCudaBase.hpp:278]: Is it necessary the extra `()` in `params.empty()` ?

File: .github/CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,9 @@
 
 ### New features since last release
 
+* Add 1-target wire controlled gate support to `lightning.tensor`. Note that `cutensornet` only supports 1-target wire controlled gate as of `v24.08`. A controlled gate with more than 1 target wires should be converted to dense matrix to get it work.
```
Comment (vincentmr) [inline .github/CHANGELOG.md:5]:
```suggestion
* Add 1-target wire controlled gate support to `lightning.tensor`. Note that `cutensornet` only supports 1-target wire controlled gate as of `v24.08`. A controlled gate with more than 1 target wire should be converted to dense matrix.
```

File: pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/TNCudaBase.hpp [hunk 3]
```diff
@@ -244,6 +246,87 @@ class TNCudaBase : public TensornetBase<PrecisionT, Derived> {
...
+     * @param controlledWires Controlled wires for the gate.
+     * @param controlled_values Controlled values for the gate.
+     * @param targetWires Target wires for the gate.
+     * @param adjoint Indicates whether to use adjoint of gate.
+     * @param params Optional parameter list for parametric gates.
+     * @param gate_matrix Optional gate matrix for custom gates.
+     */
+    void
+    applyControlledOperation(const std::string &baseOpName,
+                             const std::vector<std::size_t> &controlledWires,
```
Comment (vincentmr) [inline pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/TNCudaBase.hpp:266]: Are we using lower-camel or snake case for function arguments?
```suggestion
                             const std::vector<std::size_t> &controlled_wires,
```

**Review by vincentmr — CHANGES_REQUESTED** [review]
Looks good @multiphaseCFD . Could you run these [tests](https://github.com/PennyLaneAI/pennylane-lightning/blob/master/tests/test_gates.py#L457) with Lightning-Tensor though?

**Review by LuisAlfredoNu — APPROVED** [review]
Everything looks super good. I do not have any comment. Thank you @multiphaseCFD

**Comment by multiphaseCFD** [comment]
> Looks good @multiphaseCFD . Could you run these [tests](https://github.com/PennyLaneAI/pennylane-lightning/blob/master/tests/test_gates.py#L457) with Lightning-Tensor though?

Good point @vincentmr . Unfortunately not with current `cutensornet` API. As mentioned in the changelog, as of cutensornet v24.08, controlled gates API only support 1-wire target.

**Comment by vincentmr** [comment]
> Unfortunately not with current cutensornet API. As mentioned in the changelog, as of cutensornet v24.08, controlled gates API only support 1-wire target.

Could you add a `pytest.skip` [here](https://github.com/PennyLaneAI/pennylane-lightning/blob/master/tests/test_gates.py#L462) when `num_wires > 1`?

**Comment by multiphaseCFD** [comment]
> > Unfortunately not with current cutensornet API. As mentioned in the changelog, as of cutensornet v24.08, controlled gates API only support 1-wire target.
> 
> Could you add a `pytest.skip` [here](https://github.com/PennyLaneAI/pennylane-lightning/blob/master/tests/test_gates.py#L462) when `num_wires > 1`?

Sure, done!

**Review by vincentmr — APPROVED** [review]
LGTM, nice @multiphaseCFD .

File: .github/CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,9 @@
 
 ### New features since last release
 
+* Add 1-target wire controlled gate support to `lightning.tensor`. Note that `cutensornet` only supports 1-target wire controlled gate as of `v24.08`. A controlled gate with more than 1 target wire should be converted to dense matrix.
+  [(#879)](https://github.com/PennyLaneAI/pennylane-lightning/pull/880)
```
Comment (maliasadi) [inline .github/CHANGELOG.md:6]:
```suggestion
  [(#880)](https://github.com/PennyLaneAI/pennylane-lightning/pull/880)
```

File: pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/bindings/LTensorTNCudaBindings.hpp [hunk 5]
```diff
@@ -47,12 +48,66 @@ namespace Pennylane::LightningTensor::TNCuda {
 using TensorNetBackends =
     Pennylane::Util::TypeList<MPSTNCuda<float>, MPSTNCuda<double>, void>;
 
+/**
+ * @brief Register controlled matrix kernel.
+ */
+template <class TensorNet>
```
Comment (maliasadi) [inline pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/bindings/LTensorTNCudaBindings.hpp:54]: Do we need to use `T` for the sake of consistency?
```suggestion
template <class TensorNetT>
```

**Review by maliasadi — APPROVED** [review]
Nice work! Thanks @multiphaseCFD :rocket:

File: pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/bindings/LTensorTNCudaBindings.hpp (hunk 5 above)
Comment (maliasadi) [reply pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/bindings/LTensorTNCudaBindings.hpp:54]: Feel free to address this in the next PR.

**Pull request merged**